### PR TITLE
Fixed merge http headers listener not working with Symfony's request stack

### PR DESCRIPTION
**Pull request by Toflar**
After hours of debugging @aschempp and me finally found the issue for some redirect and cookie issues in Contao 4.4. They were introduced by using our header-replay-bundle but it was actually not the reason for the bug. The bug is that the listener instance is created once and we do not lazily fetch the headers using `headers_list()`. Now when you are using multiple request->response iterations within the same php process (which is one big advantage of Symfony and which was not noticed before because nobody did that apparently) the headers that are sent in between these iterations are ignored. We have to always fetch them lazily.

### CODE REVIEW

File: src/EventListener/MergeHttpHeadersListener.php [hunk 1]
```diff
@@ -60,13 +59,21 @@ public function __construct(ContaoFrameworkInterface $framework)
      */
     public function getHeaders()
     {
+        if ([] === $this->headers) {
+
```
Comment (leofeyer) [inline src/EventListener/MergeHttpHeadersListener.php]: This line needs to be removed. 😄

**Comment by leofeyer** [comment]
I would prefer to remove the `setHeaders()` method and use constructor injection instead.

```php
public function __construct(ContaoFrameworkInterface $framework, $headers = [])
{
    $this->headers = $headers;
}
```

File: src/EventListener/MergeHttpHeadersListener.php [hunk 2]
```diff
@@ -60,13 +59,19 @@ public function __construct(ContaoFrameworkInterface $framework)
      */
     public function getHeaders()
     {
+        if ([] === $this->headers) {
```
Comment (aschempp) [inline src/EventListener/MergeHttpHeadersListener.php]: I think `$this->headers` should be `null` if not "initialized"

Comment (leofeyer) [inline src/EventListener/MergeHttpHeadersListener.php]: That would be ok for me, too.

**Comment by Toflar** [comment]
Updated PR according to your comments.

File: src/EventListener/MergeHttpHeadersListener.php [hunk 3]
```diff
@@ -150,7 +150,7 @@ private function mergeHttpHeaders(Response $response)
     private function getHeaders()
     {
         if (null === $this->headers) {
-            return headers_list();
+            $this->headers = headers_list();
```
Comment (aschempp) [inline src/EventListener/MergeHttpHeadersListener.php]: This is wrong! Exactly the reason why we had to change the listener. The headers can change between (sub)requests, that's why we have to fetch them every time.

**Pull request merged**